Infrastructure/Services/ProcessEmailsService.cs

```diff
@@ -0,0 +1,450 @@
+            {
+                Key = TestExtraInfoKey,
+                Value = $"This is a merged email for testing purposes.{Environment.NewLine}{Environment.NewLine}"
+            });
+
+            initialEmailLog.EmailAddress = "merged_email_for_testing@education.gov.uk";
+
+            return initialEmailLog;
+        }
+
+        return null;
+    }
+
+    public bool SendTuitionPartnerEmailsWhenEnquirerDelivered()
+    {
+        return _emailSettingsConfig.SendTuitionPartnerEmailsWhenEnquirerDelivered;
+    }
+
+    private async Task<bool> StartProcessing(DateTime date)
+    {
+        var scheduledProcessingInfo = await _unitOfWork.ScheduledProcessingInfoRepository
+            .SingleOrDefaultAsync(x => x.ScheduleName == ScheduleName);
+
+        if (scheduledProcessingInfo == null)
+        {
+            _unitOfWork.ScheduledProcessingInfoRepository.AddAsync(new ScheduledProcessingInfo()
+            {
+                ScheduleName = ScheduleName,
+                LastStartedDate = date
+            });
+
+            await _unitOfWork.Complete();
+        }
+        else
+        {
+            if (scheduledProcessingInfo.LastFinishedDate == null)
+            {
+                if (scheduledProcessingInfo.LastStartedDate.AddMinutes(30) < DateTime.UtcNow)
+                {
+                    _logger.LogError("Previous ProcessAllEmailsAsync has been running for more than 30 minutes.  Was last run {LastStartedDate}, starting a fresh!", scheduledProcessingInfo.LastStartedDate);
+                }
+                else
+                {
+                    if (scheduledProcessingInfo.LastStartedDate.AddMinutes(5) < DateTime.UtcNow)
+                    {
+                        _logger.LogWarning("Previous ProcessAllEmailsAsync has been running for more than 5 minutes.  Was last run {LastStartedDate}.", scheduledProcessingInfo.LastStartedDate);
+                    }
+
+                    return false;
+                }
+            }
+
+            scheduledProcessingInfo.LastStartedDate = DateTime.UtcNow;
+            scheduledProcessingInfo.LastFinishedDate = null;
+            scheduledProcessingInfo.Status = null;
+
+            _unitOfWork.ScheduledProcessingInfoRepository.Update(scheduledProcessingInfo);
+
+            await _unitOfWork.Complete();
+        }
+
+        return true;
+    }
+
+    private async Task FinishProcessing(DateTime date, string outcomeMessage)
+    {
+        var scheduledProcessingInfo = await _unitOfWork.ScheduledProcessingInfoRepository
+            .SingleOrDefaultAsync(x => x.ScheduleName == ScheduleName);
+
+        scheduledProcessingInfo!.LastFinishedDate = date;
+        scheduledProcessingInfo!.Status = outcomeMessage;
+
+        _unitOfWork.ScheduledProcessingInfoRepository.Update(scheduledProcessingInfo);
+
+        await _unitOfWork.Complete();
+    }
+
+    private async Task PollForStatusUpdatesAsync(ProcessedEmailsModel processedEmailsModel)
```

Comment:
We should create separate methods for each function to improve the readability and maintainability of this method. We can move the `await _unitOfWork.Complete(); `out of the loop to save data to the Database after processing all emails, which will help us reduce the number of DB calls and improve performance.

Here is an example:

```
private async Task PollForStatusUpdatesAsync(ProcessedEmailsModel processedEmailsModel)
{
    var pollDateTime = DateTime.UtcNow;
    var emailsToPollForStatusUpdate = await GetEmailsToPollForStatusUpdate(pollDateTime);

    if (emailsToPollForStatusUpdate != null && emailsToPollForStatusUpdate.Any())
    {
        var emailStatuses = _unitOfWork.EmailStatusRepository.GetAll();
        Var emailsToUpdate = new List<EmailLog>();

        foreach (var emailLog in emailsToPollForStatusUpdate)
        {
            try
            {
                ValidateEmailLog(emailLog);

                processedEmailsModel.EmailsCheckStatus++;
                var currentStatus = emailLog.EmailStatus.Status.GetEnumFromDisplayName<EmailStatus>();
                var newStatus = await _notificationsClientService.GetEmailStatus(emailLog!.EmailNotifyResponseLog!.NotifyId!);

                if (currentStatus != newStatus)
                {
                    UpdateEmailLogStatus(emailLog, pollDateTime, newStatus, emailStatuses, processedEmailsModel);
                    emailsToUpdate.Add(emailLog);
                }
            }
            catch (Exception ex)
            {
                //We don't want to throw the error here, we log and continue processing other emails
                _logger.LogError(ex, "Unexpected error in emailsToPollForStatusUpdate for EmailLog.ClientReferenceNumber: {ClientReferenceNumber}, EmailLog.Id: {EmailLogId}", emailLog.ClientReferenceNumber, emailLog.Id);
            }
        }
    }

   // Update all email logs in the database at once
    foreach (var emailLog in emailsToUpdate)
    {
        _unitOfWork.EmailLogRepository.Update(emailLog);
    }

    await _unitOfWork.Complete();
}

private async Task<IEnumerable<EmailLog>> GetEmailsToPollForStatusUpdate(DateTime pollDateTime)
{
    return await _unitOfWork.EmailLogRepository
        .GetAllAsync(x => (x.EmailStatus.PollForStatusUpdateIfSent &&
                            x.LastEmailSendAttemptDate != null &&
                            x.LastEmailSendAttemptDate < pollDateTime &&
                            x.FinishProcessingDate > pollDateTime),
        "EmailStatus,EmailNotifyResponseLog,EmailsActivatedByThisEmail,EmailsActivatedByThisEmail.ActivateEmailLog,EmailsActivatedByThisEmail.ActivateEmailLog.EmailStatus",
        true);
}

private void ValidateEmailLog(EmailLog emailLog)
{
    if (emailLog == null || emailLog.EmailNotifyResponseLog == null || emailLog.EmailStatus == null)
    {
        throw new InvalidDataException("The emailsToPollForStatusUpdate has an unexpected empty data");
    }
}

private void UpdateEmailLogStatus(EmailLog emailLog, DateTime pollDateTime, EmailStatus newStatus, IEnumerable<EmailStatus> emailStatuses, ProcessedEmailsModel processedEmailsModel)
{
    LogStatusUpdate(emailLog, newStatus);

    emailLog.EmailStatusId = (int)newStatus;
    emailLog.LastStatusChangedDate = pollDateTime;

    if (newStatus == EmailStatus.NotifyDelivered)
    {
        UpdateActivatedEmailsStatus(emailLog, pollDateTime);
    }

    SetRetryDateTimeIfRequired(emailLog, pollDateTime, emailStatuses, newStatus);

    processedEmailsModel.EmailsUpdatedStatus++;
}

private void LogStatusUpdate(EmailLog emailLog, EmailStatus newStatus)
{
    var currentStatus = emailLog.EmailStatus.Status.GetEnumFromDisplayName<EmailStatus>();
    _logger.LogInformation("Email ref {ClientReferenceNumber} status update from {currentStatus}, to {newStatus}.",
        emailLog.ClientReferenceNumber,
        currentStatus.DisplayName(),
        newStatus.DisplayName());
}

private void UpdateActivatedEmailsStatus(EmailLog emailLog, DateTime pollDateTime

{
    if (emailLog!.EmailsActivatedByThisEmail != null && emailLog!.EmailsActivatedByThisEmail.Any())
    {
        _logger.LogInformation("Email ref {ClientReferenceNumber} is delivered, triggering activation of {EmailsActivatedByThisEmailCount} emails.",
            emailLog.ClientReferenceNumber,
            emailLog!.EmailsActivatedByThisEmail.Count);

        foreach (var emailToActivate in emailLog!.EmailsActivatedByThisEmail)
        {
            if (emailToActivate.ActivateEmailLog.FinishProcessingDate > pollDateTime &&
                emailToActivate.ActivateEmailLog.EmailStatus.AllowEmailSending)
            {
                emailToActivate.ActivateEmailLog.ProcessFromDate = pollDateTime;
            }
        }
    }
}

private void SetRetryDateTimeIfRequired(EmailLog emailLog, DateTime pollDateTime, IEnumerable<EmailStatus> emailStatuses, EmailStatus newStatus)
{
    var newStatusEntity = emailStatuses.Single(x => x.Id == emailLog.EmailStatusId);
    if (newStatusEntity!.RetrySendInSeconds != null)
    {
        var retryInSeconds = newStatusEntity!.RetrySendInSeconds!;
        DateTime retryDateTime = pollDateTime.AddSeconds((double)retryInSeconds);

        _logger.LogInformation("Email ref {ClientReferenceNumber} status {newStatus}, so set to retry in {retryInSeconds} seconds, on {retryDateTime}.",
            emailLog.ClientReferenceNumber,
            newStatus.DisplayName(),
            retryInSeconds,
            retryDateTime);

        emailLog.ProcessFromDate = retryDateTime;
    }
}
```